src/PowerShell/CommonFiles/PowerShellCmdlet.cs

```diff
@@ -239,21 +282,57 @@ internal void Wait(Task runningTask, PowerShellCmdlet? writeCmdlet = null)
             writeCmdlet ??= this;
 
             // This must be called in the main thread.
-            if (this.originalThread != Thread.CurrentThread)
+            if (this.pwshThread != Thread.CurrentThread)
             {
                 throw new InvalidOperationException();
             }
 
             do
             {
-                this.ConsumeAndWriteStreams(writeCmdlet);
+                if (this.pwshThreadActionReady.IsSet)
+                {
+                    // Someone needs the main thread.
+                    this.pwshThreadActionReady.Reset();
+
+                    if (this.pwshThreadAction != null)
+                    {
+                        this.pwshThreadAction();

```

Comment:
I would suggest that the try/catch be here and the exception rethrown via `WaitMainThreadActionCompletion`.